fhir-server/src/main/java/com/ibm/fhir/server/util/FHIRRestHelper.java

```diff
@@ -3087,8 +3091,123 @@ private void setOperationContextProperties(FHIROperationContext operationContext
-        String resourceLogicalId) throws Exception {
+    public int doReindex(FHIROperationContext operationContext, OperationOutcome.Builder operationOutcomeResult, Instant tstamp,
+            List<Long> indexIds, String resourceLogicalId) throws Exception {
+        int result = 0;
+        // Since the try logic is slightly different in the code paths, we want to dispatch to separate methods to simplify the logic.
+        if (indexIds == null) {
+            result = doReindexSingle(operationOutcomeResult, tstamp, resourceLogicalId);
+        } else {
+            result = doReindexList(operationOutcomeResult, tstamp, indexIds);
+        }
+        return result;
+    }
+
+    /**
+     * encapsulates the logic to process a list with graduated backoff through the full list of indexIds
+     *
+     * @param operationOutcomeResult
+     * @param tstamp
+     * @param indexIds
+     * @return
+     * @throws Exception
+     */
+    public int doReindexList(OperationOutcome.Builder operationOutcomeResult, Instant tstamp, List<Long> indexIds) throws Exception {
+        // If the indexIds are empty, assume zero.
+        if (indexIds.isEmpty()) {
+            return 0;
+        }
+
+        /*
+         * How the backoff works...
+         * indexIds[1,2,3,4,5,6,7,8,9,10]
+         *
+         * Pass 1 left=0, right=10, max=10
+         * Result: Deadlock
+         *
+         * Pass2 left=0, right=1, max=10
+         * Move over by 1
+         * Result: Pass
+         *
+         * Pass3 left=1, right=2, max=10
+         * Move over by 1
+         * Result: Pass
+         *
+         * ... all the way up to 10
+         *
+         * Return total count back to caller.
+         *
+         * @implNote tried divide and conquer and it caused it to try large pass fail, small pass succeed, and therefore chose a small linear pass.
+         */
+
+        // Maximum number to retry
+        final int TX_ATTEMPTS = 5;
+
+        int result = 0;
+        int max = indexIds.size();
+
+        int left = 0;
+        int right = max;
+
+        int window = 0;
+        while (left < max) {
+            window++;
+            if (log.isLoggable(Level.FINE)) {
+                log.fine("$reindex window[" + window + "] -> left=[" + left + "] right=[" + right + "] max=[" + max + "]");
+            }
+
+            boolean backoff = false;
+            List<Long> subListIndexIds = indexIds.subList(left, right);
+
+            int attempt = 1;
+            while (attempt <= TX_ATTEMPTS && !backoff) {
+                FHIRTransactionHelper txn = new FHIRTransactionHelper(getTransaction());
+                txn.begin();
+                try {
+                    FHIRPersistenceContext persistenceContext = null;
+                    result += persistence.reindex(persistenceContext, operationOutcomeResult, tstamp, subListIndexIds, null);
```

Comment:
I worry that the caller might interpret a non-zero as success of the whole batch. I know the contract says it's the number successfully processed...worth checking

Reply:
Right - it'll only pass up the non-zero value when successfully completed.